### PR TITLE
Constructed keyed inventory groups: Allow groups with hyphen

### LINKED ISSUE #40581: Implement replace_dash_in_groups option in aws_ec2 inventory plugin
##### SUMMARY

Implement replace_dash_in_groups in aws_ec2 inventory plugin
##### ISSUE TYPE
 - Feature Idea

##### COMPONENT NAME

aws_ec2 inventory plugin
##### ANSIBLE VERSION

```
ansible 2.5.3
  config file = /home/cam-fulcrum/ansible/ansible.cfg
  configured module search path = [u'/home/cam-fulcrum/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/cam-fulcrum/.local/share/virtualenvs/ansible-SHoMHdXE/local/lib/python2.7/site-packages/ansible
  executable location = /home/cam-fulcrum/.local/share/virtualenvs/ansible-SHoMHdXE/bin/ansible
  python version = 2.7.14 (default, Sep 23 2017, 22:06:14) [GCC 7.2.0]

```

##### CONFIGURATION

```
enable_plugins = aws_ec2, advanced_host_list
```

##### OS / ENVIRONMENT

N/A
##### STEPS TO REPRODUCE

So at the moment, if I have this in my `hosts.aws_ec2.yml`

```
plugin: aws_ec2
boto_profile: aws_profile
regions: 
  - ap-southeast-2
filters:
    instance-state-name: running
hostnames:
  - private-ip-address
strict: False
  - key: tags['Name']
    separator: ''
```

##### EXPECTED RESULTS

For instances with the name `foo-bar`, the group is turned into `foo-bar`
##### ACTUAL RESULTS

For instances with the name `foo-bar`, the group is turned into `foo_bar`
I'd like there to be a setting in the in the `hosts.aws_ec2.yml` file on whether to replace the dash with an underscore, like whats in [ec2.ini](https://github.com/ansible/ansible/blob/devel/contrib/inventory/ec2.ini#L123)

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>
Fixes #40581

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/config/base.yml
lib/ansible/plugins/inventory/__init__.py
lib/ansible/plugins/inventory/aws_ec2.py
test/integration/targets/inventory_aws_ec2/playbooks/test_populating_inventory_with_constructed.yml
test/integration/targets/inventory_aws_ec2/templates/inventory_with_constructed.yml

### CODE REVIEW

File: lib/ansible/plugins/inventory/__init__.py [hunk 1]
```diff
@@ -378,7 +378,7 @@ def _add_host_to_composed_groups(self, groups, variables, host, strict=False):
                     continue
 
                 if result:
-                    # ensure group exists, use sanatized name
+                    # ensure group exists, use santanized name
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/__init__.py:381]: sanitized

File: lib/ansible/plugins/inventory/aws_ec2.py [hunk 2]
```diff
@@ -129,6 +129,9 @@
   # create a group per region e.g. aws_region_us_east_2
   - key: placement.region
     prefix: aws_region
+  # create a group when name contains hyphens
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/aws_ec2.py:132]: To help users out, mentioning setting TRANSFORM_INVALID_GROUP_CHARS to False would be good.

File: test/integration/targets/inventory_aws_ec2/playbooks/test_populating_inventory_with_constructed.yml [hunk 3]
```diff
@@ -44,6 +44,10 @@
           set_fact:
             sg_group_name: "security_groups_{{ sg_id | replace('-', '_') }}"
 
+        - name: register the keyed and unsafe sg group name
```
Comment (s-hertel) [inline test/integration/targets/inventory_aws_ec2/playbooks/test_populating_inventory_with_constructed.yml:47]: You'll also need to set ANSIBLE_TRANSFORM_INVALID_GROUP_CHARS to False before running this

File: test/integration/targets/inventory_aws_ec2/templates/inventory_with_constructed.yml [hunk 4]
```diff
@@ -14,6 +14,9 @@ keyed_groups:
     prefix: 'tag'
   - prefix: 'arch'
     key: "architecture"
+  - key: 'unsafe_security_groups|json_query("[].group_id")'
```
Comment (s-hertel) [inline test/integration/targets/inventory_aws_ec2/templates/inventory_with_constructed.yml:17]: Right now the key (i.e. host var) is `security_groups`. Maybe you meant
```
key: 'security_groups | json_query("[].group_id")'
prefix: 'unsafe_security_groups'
unsafe: True
```

File: lib/ansible/plugins/inventory/__init__.py [hunk 5]
```diff
@@ -415,7 +415,11 @@ def _add_host_to_keyed_groups(self, keys, variables, host, strict=False):
                             raise AnsibleParserError("Invalid group name format, expected a string or a list of them or dictionary, got: %s" % type(key))
 
                         for bare_name in new_raw_group_names:
-                            gname = to_safe_group_name('%s%s%s' % (prefix, sep, bare_name))
+                            if keyed.get('unsafe', False):
+                                gname = '%s%s%s' % (prefix, sep, bare_name)
```
Comment (s-hertel) [inline lib/ansible/plugins/inventory/__init__.py:419]: Right now I get conflicting warnings about the group replacing invalid characters and then that the group doesn't exist (and then inventory fails to parse) when not setting the global toggle but using `unsafe`. `force` and `silent` should probably become parameters of that function on line 42 so you can call that instead.

**Comment by AlanCoding** [comment]
Thanks for the PR! I'm still mulling over all the content, so I hope you don't mind if I leave some only slightly related notes here.

the `ec2.py` script has a special case, which @bcoca has pointed out to me

https://github.com/ansible/ansible/blob/874fd70d103c1bd70de2a9ea6523d447a861f486/contrib/inventory/ec2.py#L1692-L1697

I have been racking my brain to find all differences between this and the general on/off toggle, and I think it's just `-`. That's all that I know of.

The general sanitization is done at:

https://github.com/ansible/ansible/blob/d241794daa6d413e6447890e2a4f11e0d818cf0e/lib/ansible/constants.py#L122

Again, I think `^[^a-zA-Z_]|[^a-zA-Z0-9_]` == `[^A-Za-z0-9\_]`, but I am not a regex whiz by any means.

Azure does

https://github.com/ansible/ansible/blob/874fd70d103c1bd70de2a9ea6523d447a861f486/contrib/inventory/azure_rm.py#L963-L968

...looks like they copied from ec2.

In gce.py, I can't find any name sanitization from the script.

https://github.com/ansible/ansible/blob/874fd70d103c1bd70de2a9ea6523d447a861f486/contrib/inventory/gce.py

Now, for the plugins, all 3 of these go through this keyed_groups thing that you're editing.

For all of ec2, azure, and gce, this patch should allow us to restore the old group names for all groups that don't have `-` in them, or if the replace_dashes setting is turned off (again, I am no good at regex so correct me if I'm wrong).

This patch will not solve the case of reproducing the replace_dashes=true behavior for ec2 and azure.

**Comment by AlanCoding** [comment]
This still might be needed - I want to obtain behavior of the old ec2.py script, and that script selectively did not sanitize region names. I could do the same with this tool.

**Comment by AlanCoding** [comment]
Note to myself, I did reopen https://github.com/ansible/ansible/pull/51958 which does the same thing as this, it will need to be closed once it's confirmed that one of the alternatives is merged.

**Comment by babbottscott** [comment]
side note re:
> Again, I think ^[^a-zA-Z_]|[^a-zA-Z0-9_] == [^A-Za-z0-9\_], but I am not a regex whiz by any means.

The former does not allow a leading numeral.  The latter does.

**Pull request closed without merging**